geomstats/stratified_geometry/wald_space.py

```diff
@@ -0,0 +1,1024 @@
+            A float between 0 and 1 determining the probability with which a split
+            is kept and not deleted.
+        check : bool
+            If True, checks if splits still separate all labels. In this case, the split
+            will not be deleted. If False, any split can be randomly deleted.
+
+        Returns
+        -------
+        left_over_splits : list[Split]
+            The list of splits that are not deleted.
+        """
+        if p_keep == 1:
+            return splits
+        for i in reversed(range(len(splits))):
+            if gs.random.rand(1) > p_keep:
+                splits_cp = splits.copy()
+                splits_cp.pop(i)
+                if not check:
+                    splits = splits_cp
+                elif Wald._check_if_separated(splits=splits_cp, labels=labels):
+                    splits = splits_cp
+        return splits
+
+    @staticmethod
+    def generate_wald(n, p_keep, p_new, btol=10**-8, check=True):
+        """Generate a random instance of class ``Wald``.
+
+        Parameters
+        ----------
+        n : int
+            The number of labels the wald is generated with respect to.
+        p_keep : float
+            The probability will be inserted into the generation of a partition as
+            well as for the generation of a split set for the topology of the wald.
+        p_new : float
+            A float between 0 and 1, the probability that no new component is added,
+            and probability of 1 - p_new_ that a new component is added.
+        btol: float
+            Tolerance for the boundary of the coordinates in each grove. Defaults to
+            1e-08.
+        check : bool
+            If True, checks if splits still separate all labels. In this case, the split
+            will not be deleted. If False, any split can be randomly deleted.
+
+        Returns
+        -------
+        random_wald : Wald
+            The randomly generated wald.
+        """
+        partition = Wald._generate_partition(n=n, p_new=p_new)
+        split_sets = [Wald._generate_splits(labels=_part) for _part in partition]
+        split_sets = [
+            Wald._delete_splits(splits=splits, labels=part, p_keep=p_keep, check=check)
+            for part, splits in zip(partition, split_sets)
+        ]
+        top = Topology(n=n, partition=partition, split_sets=split_sets)
+        x = gs.random.uniform(size=(len(top.flatten(split_sets)),), low=0, high=1)
+        x = np.minimum(np.maximum(btol, x), 1 - btol)
+        return Wald(n=n, st=top, x=x)
+
+
+class WaldSpace(PointSet):
+    """Class for the Wald space, a metric space for phylogenetic forests.
+
+    Parameters
+    ----------
+    n : int
+        Integer determining the number of labels in the forests, and thus the shape of
+        the correlation matrices: n x n.
+    """
+
+    def __init__(self, n):
+        super(WaldSpace).__init__()
+        self.n = n
+        self.a = spd.SPDMatrices(n=self.n)
```

Comment:
What is `self.a`? Again, naming. Attributes of the class can also go to the class docstrings (under `Attributes`).

Reply:
I'll call it self.ambient now, that's ok